api/server/runner_fninvoke.go

```diff
@@ -66,31 +78,35 @@ func (s *Server) handleFnInvokeCall2(c *gin.Context) error {
 }
 
 func (s *Server) ServeFnInvoke(c *gin.Context, app *models.App, fn *models.Fn) error {
-	// TODO: we should combine this logic with trigger, which just wraps this block with some headers wizardry
 	// TODO: we should get rid of the buffers, and stream back (saves memory (+splice), faster (splice), allows streaming, don't have to cap resp size)
+	//TODO: Remove this pool, it makes concurrency bugs easy to create, but hard to discover.
 	buf := bufPool.Get().(*bytes.Buffer)
 	buf.Reset()
-	writer := syncResponseWriter{
+	defer bufPool.Put(buf) // TODO need to ensure this is safe with Dispatch?
+
+	writer := &syncResponseWriter{
 		Buffer:  buf,
 		headers: c.Writer.Header(),
 	}
-	defer bufPool.Put(buf) // TODO need to ensure this is safe with Dispatch?
 
-	call, err := s.agent.GetCall(
-		agent.WithWriter(&writer), // XXX (reed): order matters [for now]
-		agent.FromHTTPFnRequest(app, fn, c.Request),
-	)
+	call, err := s.agent.GetCall(agent.WithWriter(writer), // XXX (reed): order matters [for now]
+		agent.FromHTTPFnRequest(app, fn, c.Request))
+
 	if err != nil {
 		return err
 	}
 
+	return s.FnInvoke(c, app, fn, writer, call)
+}
+
+func (s *Server) FnInvoke(c *gin.Context, app *models.App, fn *models.Fn, writer ResponseBufferingWriter, call agent.Call) error {
```

Comment:
this shouldn't be exported, it has a very contextually specific contract and the parameters don't make sense from an exported API (nor does ResponseBufferingWriter need be exported) - please lower case